### PR TITLE
RavenDB-7503

### PR DESCRIPTION
**Pull request by talweiss1982**
* Now allowing api keys with '/' as long as they are starting with Raven/ and properly split their secret.
* Opening an async task listening to a named pipe constructed as raven-control-pipe-<raven server pid>
* Added a method to create a system api key named Raven/System to be used by the admin in harden environment
* Added trust/init methods to RVN tool, still need TLC will open another issue to have better error handling and validation in RVN

### CODE REVIEW

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 1]
```diff
@@ -1395,12 +1398,78 @@ public DynamicJsonValue GetLogDetails(TransactionOperationContext context, int m
 
         }
 
+        private const string SystemApiKey = "Raven/System";
+        public async Task<(string api_key, string public_key)> GetApiKeyAndPublicKey()
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ServerStore.cs]: @talweiss1982 can you use C# casing here? Task<(string ApiKey, string PublicKey)>

File: src/Raven.Server/RavenServer.cs [hunk 2]
```diff
@@ -119,7 +121,10 @@ public void Initialize()
                 _logger.Info(string.Format("Server store started took {0:#,#;;0} ms", sp.ElapsedMilliseconds));
 
             sp.Restart();
-
+            if (Pipe != null)
+            {
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Add comment explaining why this is.
This is the second time I'm seeing this code and going `HuGH?`

File: src/Raven.Server/RavenServer.cs [hunk 3]
```diff
@@ -184,6 +189,95 @@ public void Initialize()
             }
         }
 
+        private async Task ListenToPipe()
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Missing error handling here.

File: src/Raven.Server/RavenServer.cs [hunk 4]
```diff
@@ -184,6 +189,95 @@ public void Initialize()
...
+                }
+                catch (Exception e)
+                {
+                    if (_logger.IsInfoEnabled)
+                    {
+                        _logger.Info("Got an exception trying to connect to server pipe", e);
+                    }
+                }
+                //From what i read we should not re-use a pipe.
+                Pipe.Dispose();
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Failure to open / dispose would crash the process

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 1 above)
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: Please use PascalCase here.

File: src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs [hunk 5]
```diff
@@ -176,7 +176,13 @@ private string[] ExtractApiKeyAndSecret(string apiKey)
             var apiKeyParts = apiKey.Split(new[] { '/' }, StringSplitOptions.None);
             if (apiKeyParts.Length > 2)
             {
-                apiKeyParts[1] = string.Join("/", apiKeyParts.Skip(1));
+                var secretIndex = 1;
+                if (apiKeyParts[0] == "Raven" && apiKeyParts.Length > 3)
```
Comment (ayende) [inline src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs]: Maybe better of to use "Raven:Cluster/Secret" ?

Comment (ayende) [reply src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs]: And then reserve just `Raven:`

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs (hunk 5 above)
Comment (ppekrol) [reply src/Raven.Client/Http/OAuth/ApiKeyAuthenticator.cs]: I think that 'Raven/' is already reserved: https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Web/Authentication/AdminApiKeysHandler.cs#L26

File: src/Raven.Server/RavenServer.cs [hunk 6]
```diff
@@ -189,82 +190,96 @@ public void Initialize()
             }
         }
 
-        private async Task ListenToPipe()
+        internal async Task ListenToPipe()
         {
-            while (true)
+            lock (this)
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: What, why? This is going to be locking effectively forever.

File: src/Raven.Server/Web/Authentication/AdminApiKeysHandler.cs [hunk 7]
```diff
@@ -68,6 +68,13 @@ public class AdminApiKeysHandler : AdminRequestHandler
             }
         }
 
+        [RavenAction("/admin/server/reset-pipe", "GET", "/admin/server/reset-pipe")]
```
Comment (ayende) [inline src/Raven.Server/Web/Authentication/AdminApiKeysHandler.cs:71]: Why do we need this?

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ayende** [comment]
We are going with a different route here, see: http://issues.hibernatingrhinos.com/issue/RavenDB-7696

**Pull request closed without merging**